src/aspire/source/relion.py

```diff
@@ -103,48 +119,100 @@ def __init__(
+                        Cs=row[4],
+                        alpha=row[5],
+                        B=B,
+                    )
+                )
+            self.unique_filters = filters
+            # filter_indices stores, for each particle index, the index in
+            # self.unique_filters of the filter that should be applied
+            self.filter_indices = filter_indices
+        # If no CTF info in STAR, we initialize the filter values of metadata with default values
+        else:
+            self.unique_filters = [IdentityFilter()]
+            self.filter_indices = np.zeros(self.n, dtype=int)
 
-        ImageSource.__init__(
-            self, L=L, n=n, dtype=dtype, metadata=metadata, memory=memory
+    def populate_metadata(self, filepath, data_folder=None, max_rows=None):
+        """
+        Relion STAR files may contain a large number of metadata columns in addition
+        to the locations of particles. We read this into a Pandas DataFrame and add some of
+        our own columns for convenience.
+        """
+        if data_folder is not None:
+            if not os.path.isabs(data_folder):
+                data_folder = os.path.join(os.path.dirname(filepath), data_folder)
+        else:
+            data_folder = os.path.dirname(filepath)
+
+        # Valid Relion STAR files always have their data in the first loop of the first block.
+        # We are getting the first (and only) block in this StarFile object
+        df = StarFile(filepath).get_block_by_index(0)
+        # convert STAR file strings to data type for each field
+        # columns without a specified data type are read as dtype=object
+        column_types = {
+            name: RelionSource.relion_metadata_fields.get(name, str)
+            for name in df.columns
+        }
+        df = df.astype(column_types)
+
+        # particle locations are stored as e.g. '000001@first_micrograph.mrcs'
+        # in the _rlnImageName column. here, we're splitting this information
+        # so we can get the particle's index in the .mrcs stack as an int
+        df[["__mrc_index", "__mrc_filename"]] = df["_rlnImageName"].str.split(
+            "@", 1, expand=True
         )
-        self.unique_filters = filters
-        self.filter_indices = filter_indices
+        # __mrc_index corresponds to the integer index of the particle in the __mrc_filename stack
+        # Note that this is 1-based indexing
+        df["__mrc_index"] = pd.to_numeric(df["__mrc_index"])
+
+        # Adding a full-filepath field to the Dataframe helps us save time later
+        # Note that os.path.join works as expected when the second argument is an absolute path itself
+        df["__mrc_filepath"] = df["__mrc_filename"].apply(
+            lambda filename: os.path.join(data_folder, filename)
+        )
+
+        if max_rows is None:
+            return df
+        else:
+            max_rows = min(max_rows, len(df))
+            return df.iloc[:max_rows]
```

Comment:
Not sure if it matters here, but offhand doesn't pandas loc/iloc potentially return view? I remember fixing some bugs like that. Not saying this is wrong, but just double check this does exactly what you want.

Reply:
Double checked that this returns a DF when given a slice like this. Thanks!